lib/classes/ConfigSchemaHandler.js

```diff
@@ -0,0 +1,143 @@
+'use strict';
+
+const Ajv = require('ajv');
+const _ = require('lodash');
+const schema = require('../configSchema');
+const errorMessageBuilder = require('./ConfigSchemaHandler/errorMessageBuilder');
+
+const FUNCTION_NAME_PATTERN = '^[a-zA-Z0-9-_]+$';
+const ERROR_PREFIX = 'Configuration error:';
+
+class ConfigSchemaHandler {
+  constructor(serverless) {
+    this.serverless = serverless;
+    this.schema = _.cloneDeep(schema);
+
+    deepFreeze(this.schema.properties.service);
+    deepFreeze(this.schema.properties.plugins);
+    deepFreeze(this.schema.properties.package);
+    Object.freeze(this.schema.properties.layers);
+    Object.freeze(this.schema.properties.resources);
+  }
+
+  validateConfig(userConfig) {
+    const ajv = new Ajv();
+    const validate = ajv.compile(this.schema);
+
+    // console.log('userConfig', require('util').inspect(userConfig, false, null, true));
+    // console.log('=====');
+    // console.log('schema', require('util').inspect(this.schema, false, null, true));
+
+    // In case provider name is not 'aws' and no other provider schema was introduced,
+    // then this means that no other provider plugin was loaded. In this case
+    // we need to throw an error.
+    if (
+      this.serverless.service.provider.name !== 'aws' &&
+      this.schema.properties.provider.properties.name === undefined
+    ) {
+      this.handleErrorMessages([
+        `Unsupported provider name '${this.serverless.service.provider.name}'`,
+      ]);
+    }
+
+    validate(userConfig);
+    if (validate.errors) {
+      // console.log('ajv errors', require('util').inspect(validate.errors, false, null, true));
+      const messages = errorMessageBuilder.buildErrorMessages(validate.errors, userConfig);
+      this.handleErrorMessages(messages);
+    }
+  }
+
+  handleErrorMessages(messages) {
+    for (const message of messages) {
+      switch (this.serverless.service.configValidationMode) {
+        case 'error':
+          throw new this.serverless.classes.Error(`${ERROR_PREFIX} ${messages.join('; ')}`);
+        case 'warn':
+          this.serverless.cli.log(`${ERROR_PREFIX} ${message}`, 'Serverless', { color: 'red' });
+          break;
+        default:
+          throw new this.serverless.classes.Error(
+            `${ERROR_PREFIX} .configValidationMode should be equal to one of the allowed values 'error' or 'warn'`
```

Comment:
I once tried to make it cleaner but faced various issues of improper behavior. I believe `configValidationMode` should be treated a bit special (as well as `aws` provider name) because the validation logic depends on it. I'd say that things that define the system cannot be handled by the system.

I'll remove the leading dot `.` however

Reply:
> faced various issues of improper behavior.

What issues exactly? As I look in it, there seems nothing special about it

Reply:
@medikoo here's one issue: https://github.com/serverless/serverless/pull/7335#discussion_r414341921

Reply:
If we'd do as I proposed here: https://github.com/serverless/serverless/pull/7335#discussion_r437892047 I don't think it'll throw (it's `warn` that should be seen as _default_ not `error`)

Reply:
Thanks for the clarification. I've updated it in https://github.com/serverless/serverless/pull/7335/commits/0fee18c8fd205f720885e9dfb4e674e779830fe2 and this seems to properly